File: Java/maven/src/main/java/com/automatak/dnp3/impl/CommandHandlerAdapter.java
package com.automatak.dnp3.impl;

import com.automatak.dnp3.*;

// Makes the JNI callbacks easier by breaking down the objects and passing enums as ints
class CommandHandlerAdapter {

    private final CommandHandler proxy;

    public CommandHandlerAdapter(CommandHandler proxy)
    {
        this.proxy = proxy;
    }

    // ControlRelayOutputBlock

    public int Select(int function, short count, long onTimeMs, long offTimeMs, int status, long index, byte aSeq)
    {
        ControlRelayOutputBlock crob = new ControlRelayOutputBlock(ControlCode.fromInt(function), count, onTimeMs, offTimeMs, CommandStatus.fromInt(status));
        return proxy.Select(crob, index).toInt();
    }

    public int Operate(int function, short count, long onTimeMs, long offTimeMs, int status, long index, byte aSeq)
    {
        ControlRelayOutputBlock crob = new ControlRelayOutputBlock(ControlCode.fromInt(function), count, onTimeMs, offTimeMs, CommandStatus.fromInt(status));
        return proxy.Operate(crob, index).toInt();
    }

    public int DirectOperate(int function, short count, long onTimeMs, long offTimeMs, int status, long index)
    {
        ControlRelayOutputBlock crob = new ControlRelayOutputBlock(ControlCode.fromInt(function), count, onTimeMs, offTimeMs, CommandStatus.fromInt(status));
        return proxy.DirectOperate(crob, index).toInt();
    }

    // AnalogOutputInt32

    public int Select(int value, long index, byte aSeq)
    {
        AnalogOutputInt32 ao = new AnalogOutputInt32(value, CommandStatus.SUCCESS);
        return proxy.Select(ao, index).toInt();
    }

    public int Operate(int value, long index, byte aSeq)
    {
        AnalogOutputInt32 ao = new AnalogOutputInt32(value, CommandStatus.SUCCESS);
        return proxy.Operate(ao, index).toInt();
    }

    public int DirectOperate(int value, long index)
    {
        AnalogOutputInt32 ao = new AnalogOutputInt32(value, CommandStatus.SUCCESS);
        return proxy.DirectOperate(ao, index).toInt();
    }

    // AnalogOutputInt16

    public int Select(short value, long index, byte aSeq)
    {
        AnalogOutputInt16 ao = new AnalogOutputInt16(value, CommandStatus.SUCCESS);
        return proxy.Select(ao, index).toInt();
    }

    public int Operate(short value, long index, byte aSeq)
    {
        AnalogOutputInt16 ao = new AnalogOutputInt16(value, CommandStatus.SUCCESS);
        return proxy.Operate(ao, index).toInt();
    }

    public int DirectOperate(short value, long index)
    {
        AnalogOutputInt16 ao = new AnalogOutputInt16(value, CommandStatus.SUCCESS);
        return proxy.DirectOperate(ao, index).toInt();
    }

    // AnalogOutputFloat32

    public int Select(float value, long index, byte aSeq)
    {
        AnalogOutputFloat32 ao = new AnalogOutputFloat32(value, CommandStatus.SUCCESS);
        return proxy.Select(ao, index).toInt();
    }

    public int Operate(float value, long index, byte aSeq)
    {
        AnalogOutputFloat32 ao = new AnalogOutputFloat32(value, CommandStatus.SUCCESS);
        return proxy.Operate(ao, index).toInt();
    }

    public int DirectOperate(float value, long index)
    {
        AnalogOutputFloat32 ao = new AnalogOutputFloat32(value, CommandStatus.SUCCESS);
        return proxy.DirectOperate(ao, index).toInt();
    }

    // AnalogOutputDouble64

    public int Select(double value, long index, byte aSeq)
    {
        AnalogOutputDouble64 ao = new AnalogOutputDouble64(value, CommandStatus.SUCCESS);
        return proxy.Select(ao, index).toInt();
    }

    public int Operate(double value, long index, byte aSeq)
    {
        AnalogOutputDouble64 ao = new AnalogOutputDouble64(value, CommandStatus.SUCCESS);
        return proxy.Operate(ao, index).toInt();
    }

    public int DirectOperate(double value, long index)
    {
        AnalogOutputDouble64 ao = new AnalogOutputDouble64(value, CommandStatus.SUCCESS);
        return proxy.DirectOperate(ao, index).toInt();
    }

}

File: Java/cpp/CommandHandlerAdapter.cpp

#include "CommandHandlerAdapter.h"
#include "JNIHelpers.h"

using namespace apl::dnp;

CommandHandlerAdapter::CommandHandlerAdapter(JavaVM* apJVM, jobject aProxy) : mpJVM(apJVM), mProxy(aProxy)
{}

CommandStatus CommandHandlerAdapter::Select(const ControlRelayOutputBlock& arCommand, size_t aIndex, uint8_t aSequence)
{
	return CS_SUCCESS;
/*
	auto pEnv = GetEnv();
	jclass clazz = pEnv->GetObjectClass(mProxy);
	assert(clazz != NULL);
	jmethodid mid = pEnv->GetMethodId(clazz, "Select", "(ISJJIJ)I");
	assert(mid != NULL);
	//int function, short count, long onTimeMs, long offTimeMs, int status, long index)
	jint func = arCommand
	jint status = pEnv->CallIntMethod(mProxy, mid, arCommand.mFuncit 
	*/
}

CommandStatus CommandHandlerAdapter::Operate(const ControlRelayOutputBlock& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::DirectOperate(const ControlRelayOutputBlock& arCommand, size_t aIndex)
{ return CS_SUCCESS; }

CommandStatus CommandHandlerAdapter::Select(const AnalogOutputInt16& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::Operate(const AnalogOutputInt16& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::DirectOperate(const AnalogOutputInt16& arCommand, size_t aIndex)
{ return CS_SUCCESS; }

CommandStatus CommandHandlerAdapter::Select(const AnalogOutputInt32& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::Operate(const AnalogOutputInt32& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::DirectOperate(const AnalogOutputInt32& arCommand, size_t aIndex)
{ return CS_SUCCESS; }

CommandStatus CommandHandlerAdapter::Select(const AnalogOutputFloat32& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::Operate(const AnalogOutputFloat32& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::DirectOperate(const AnalogOutputFloat32& arCommand, size_t aIndex)
{ return CS_SUCCESS; }

CommandStatus CommandHandlerAdapter::Select(const AnalogOutputDouble64& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::Operate(const AnalogOutputDouble64& arCommand, size_t aIndex, uint8_t aSequence)
{ return CS_SUCCESS; }
CommandStatus CommandHandlerAdapter::DirectOperate(const AnalogOutputDouble64& arCommand, size_t aIndex)
{ return CS_SUCCESS; }
	
JNIEnv* CommandHandlerAdapter::GetEnv()
{
	return JNIHelpers::GetEnvFromJVM(mpJVM);
}





